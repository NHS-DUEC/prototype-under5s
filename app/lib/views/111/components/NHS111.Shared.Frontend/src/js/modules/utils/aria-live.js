export function ariaAssertiveMessage(message) {
  const ariaAssertiveRegion = document.getElementById("aria-assertive");
  const el = document.createElement("p");
  el.textContent = message;
  ariaAssertiveRegion.appendChild(el);

  setTimeout(() => {
    el.remove();
  }, 100);
}

export function ariaPoliteMessage(message) {
  const ariaPoliteRegion = document.getElementById("aria-polite");
  const el = document.createElement("p");
  el.textContent = message;
  ariaPoliteRegion.appendChild(el);

  setTimeout(() => {
    el.remove();
  }, 100);
}
