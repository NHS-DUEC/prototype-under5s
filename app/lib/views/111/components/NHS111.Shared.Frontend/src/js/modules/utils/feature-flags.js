const featureFlags = JSON.parse(
  document.querySelector("#feature-flags").innerHTML
);

export function featureIsEnabled(featureName) {
  return featureFlags[featureName];
}
