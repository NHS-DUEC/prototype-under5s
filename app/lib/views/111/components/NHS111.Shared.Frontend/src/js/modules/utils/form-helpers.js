export function hideNoJsMessage(form) {
  Array.from(form.querySelectorAll("[data-feedback-nojs-message]")).forEach(
    (item) => item.remove()
  );
}

export function clearErrors(form) {
  const errors = form.querySelectorAll("[data-feedback-error]");
  const formGroups = form.querySelectorAll(".nhsuk-form-group");
  const elementsWithOriginalAria = form.querySelectorAll(
    "[data-original-aria-describedby]"
  );

  Array.from(errors).forEach((item) => item && item.remove());
  Array.from(formGroups).forEach(
    (item) => item && item.classList.remove("nhsuk-form-group--error")
  );
  Array.from(elementsWithOriginalAria).forEach((el) =>
    el.setAttribute("aria-describedBy", el.dataset.originalAriaDescribedby)
  );
}

export function hideAllFormSections(form, formSections) {
  clearErrors(form);

  Array.from(formSections).forEach((element) => {
    element.hidden = true;
  });
}

export function trackOriginalAria(form) {
  // Keep track of any original aria-describedBy attributes so we can restore them when clearing errors
  Array.from(form.querySelectorAll("fieldset, input, textarea")).forEach(
    (el) => {
      if (el.getAttribute("aria-describedBy") !== null) {
        el.setAttribute(
          "data-original-aria-describedby",
          el.getAttribute("aria-describedBy")
        );
      }
    }
  );
}

export function formError(
  form,
  ErrorSummary,
  fieldId,
  errorMessage,
  errorLink = ""
) {
  const template = document.createElement("template");

  clearErrors(form);

  template.innerHTML = `
    <div class="nhsuk-error-summary" data-feedback-error data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#${fieldId}">${errorLink || errorMessage}</a>
          </li>
        </ul>
      </div>
    </div>`;

  const previousErrorSummary = form.querySelector(".nhsuk-error-summary");
  if (previousErrorSummary) {
    previousErrorSummary.remove();
  }

  const errorSummary = template.content;

  form.prepend(errorSummary);
  ErrorSummary();

  const field = form.querySelector(`#${fieldId}`);
  const formGroup = field.closest(".nhsuk-form-group");

  const errorMessageTemplate = document.createElement("template");
  errorMessageTemplate.innerHTML = `<span data-feedback-error data-test-id="feedback-error-message" class="nhsuk-error-message" id="${fieldId}-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> ${errorMessage}
    </span>`;

  if (field.getAttribute("type") === "radio") {
    formGroup.querySelector("legend").after(errorMessageTemplate.content);
    const fieldset = formGroup.querySelector("fieldset");
    fieldset.setAttribute(
      "aria-describedby",
      `${
        fieldset.dataset.originalAriaDescribedby
          ? fieldset.dataset.originalAriaDescribedby
          : ""
      } ${fieldId}-error`
    );
  } else {
    formGroup.querySelector("label").after(errorMessageTemplate.content);
    field.setAttribute(
      "aria-describedby",
      `${
        field.dataset.originalAriaDescribedby
          ? field.dataset.originalAriaDescribedby
          : ""
      } ${fieldId}-error`
    );
  }

  formGroup.classList.add("nhsuk-form-group--error");
}
