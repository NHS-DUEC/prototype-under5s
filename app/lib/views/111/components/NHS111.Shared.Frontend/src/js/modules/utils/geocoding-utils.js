const addressStringToLatLongCache = {};

export async function addressStringToLatLong(
  GoogleMapsGeocoder,
  addressString
) {
  return new Promise((resolve, reject) => {
    if (addressString in addressStringToLatLongCache) {
      resolve(addressStringToLatLongCache[addressString]);
    } else {
      const geocoder = new GoogleMapsGeocoder();
      geocoder.geocode(
        { address: addressString, componentRestrictions: { country: "GB" } },
        (data) => {
          addressStringToLatLongCache[addressString] =
            data[0].geometry.location;
          resolve(addressStringToLatLongCache[addressString]);
        }
      );
    }
  });
}
