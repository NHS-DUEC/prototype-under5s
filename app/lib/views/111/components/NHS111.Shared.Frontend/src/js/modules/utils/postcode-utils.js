function lastIndexOfDigit(str) {
  const lastIndices = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map((digit) =>
    str.lastIndexOf(digit)
  );
  const indices = lastIndices.sort((a, b) => {
    return a - b;
  });
  return indices[indices.length - 1];
}

export function getPartPostCode(postcode) {
  const postcodeWithoutWhitespace = postcode.replace(/\s/g, "");

  if (postcodeWithoutWhitespace.length - 3 <= 1)
    return postcodeWithoutWhitespace;

  const lastDigit = lastIndexOfDigit(postcodeWithoutWhitespace);

  if (lastDigit === -1) {
    return postcodeWithoutWhitespace;
  }

  if (lastDigit < postcodeWithoutWhitespace.length - 1)
    return postcodeWithoutWhitespace.substring(0, lastDigit);

  return postcodeWithoutWhitespace;
}
