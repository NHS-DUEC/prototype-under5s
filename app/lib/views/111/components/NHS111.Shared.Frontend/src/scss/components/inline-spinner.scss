.app-inline-spinner {
  color: $color_nhsuk-white;
  display: block;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 30px;
  height: 30px;
  margin: 0 !important; // Because of the lobotomised owl. Get rid of this when we can
}

.app-inline-spinner span {
  transform-origin: 15px 15px;
  animation: app-inline-spinner 1.2s linear infinite;
  display: block;
  margin: 0 !important; // Because of the lobotomised owl. Get rid of this when we can
}

.app-inline-spinner span:after {
  content: " ";
  display: block;
  position: absolute;
  top: 2px;
  left: 13px;
  width: 4px;
  height: 8px;
  border-radius: 20%;
  background: $color_nhsuk-white;
}

@for $i from 0 through 11 {
  .app-inline-spinner span:nth-child(#{$i + 1}) {
    transform: rotate(0deg + 30 * $i);
    animation-delay: -1.1s + $i * 0.1;
  }
}

@keyframes app-inline-spinner {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}
