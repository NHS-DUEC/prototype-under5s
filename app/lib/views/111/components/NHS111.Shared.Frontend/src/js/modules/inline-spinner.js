import { ariaPoliteMessage } from "./utils/aria-live";

export class InlineSpinner {
  constructor(el) {
    this.el = el;
    this.intercept = true;
    this.originalText = this.el.textContent;
    this.isDoubleClickPrevented = !!this.el.dataset.preventDoubleClick;
    this.spinner = document.createElement("span");
    this.spinner.innerHTML = `<span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span>`;
    this.spinner.classList.add("app-inline-spinner");
  }

  showInlineSpinner() {
    ariaPoliteMessage(`Loading`);
    this.el.setAttribute("disabled", "");
    this.el.textContent = "Please wait";
    this.el.appendChild(this.spinner);
  }

  hideLoadingSpinner() {
    this.spinner.remove();
    this.el.removeAttribute("disabled");
    this.el.textContent = this.originalText;
  }

  shouldShowSpinner() {
    // Based on the data-inline-spinner-condition attribute - look for elements that meet the criteria and only show spinner if they exist
    // This is used on question pages to only show the spinner if a radio button with a nextNodeLabel of outcome is checked
    if (this.el.dataset.inlineSpinnerCondition) {
      return !!document.querySelector(this.el.dataset.inlineSpinnerCondition);
    }

    return true;
  }

  handleEvent(e) {
    if (this.shouldShowSpinner() && this.intercept) {
      e.preventDefault();
      this.showInlineSpinner();

      // Fire off the original event, but make sure we don't intercept it again
      this.intercept = false;

      setTimeout(() => {
        this.el.removeAttribute("disabled");

        if (this.isDoubleClickPrevented) {
          this.el.removeAttribute("data-prevent-double-click");
        }

        this.el.click();
        this.el.setAttribute("disabled", "");

        if (this.isDoubleClickPrevented) {
          this.el.setAttribute("data-prevent-double-click", "true");
        }

        setTimeout(this.hideLoadingSpinner.bind(this), 30000); // Hide the loading spinner again after 30s as a safety net - let people try again

        this.intercept = true;
      }, 100); // Non zero delay to allow voiceover time to register the aria message
    }
  }

  init() {
    if (this.el.form) {
      this.el.form.addEventListener("submit", this.handleEvent.bind(this));
    } else {
      this.el.addEventListener("click", this.handleEvent.bind(this));
    }

    window.addEventListener("pagehide", this.hideLoadingSpinner.bind(this));
    window.addEventListener("pageshow", this.hideLoadingSpinner.bind(this));
  }
}

export function initInlineSpinners() {
  // eslint-disable-next-line no-prototype-builtins
  if (!Element.prototype.hasOwnProperty("remove")) {
    // IE11 can't manage this so bail out
    return;
  }

  const inlineSpinnerTriggers = Array.from(
    document.querySelectorAll("[data-inline-spinner]")
  );
  inlineSpinnerTriggers.forEach((el) => {
    new InlineSpinner(el).init();
  });
}
