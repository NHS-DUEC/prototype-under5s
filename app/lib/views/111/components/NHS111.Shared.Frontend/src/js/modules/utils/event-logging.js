import { getPartPostCode } from "./postcode-utils";

export function getHiddenFieldValue(name, nameEndsWith = false) {
  const fields = nameEndsWith
    ? document.querySelectorAll(`[name$="${name}"]`)
    : document.getElementsByName(name);

  if (fields.length > 0) {
    return fields[0].value;
  }

  return "";
}

export function getGender() {
  const gender = getHiddenFieldValue("Gender", true);

  if (["male", "female"].includes(gender.toLowerCase())) return gender;

  return "";
}

export function getDxCode() {
  // find the DxCode by searching for any input with value starting Dx
  const dxCode = document.querySelector('input[value^="Dx"]');

  if (dxCode) {
    return dxCode.value;
  }

  return "";
}

export function createAuditEntry(
  eventKey,
  eventValue,
  eventData = "",
  answerOrder = "",
  answerTitle = ""
) {
  return {
    journeyId: getHiddenFieldValue("JourneyId", true),
    campaign: getHiddenFieldValue("Campaign"),
    dxCode: getDxCode(),
    postCodePart: getPartPostCode(getHiddenFieldValue("CurrentPostcode", true)),
    pathwayId: getHiddenFieldValue("PathwayId", true),
    pathwayTitle: getHiddenFieldValue("PathwayTitle", true),
    campaignSource: getHiddenFieldValue("Source"),
    QuestionNo: getHiddenFieldValue("QuestionNo"),
    QuestionTitle: getHiddenFieldValue("Title"),
    age: getHiddenFieldValue("Age", true),
    gender: getGender(),
    eventKey,
    eventValue,
    eventData,
    page: window.location.pathname,
    AnswerOrder: answerOrder,
    AnswerTitle: answerTitle,
  };
}

export function logEvent(
  eventKey,
  eventValue,
  eventData,
  answerOrder,
  answerTitle
) {
  const formData = new FormData();
  const auditEntry = createAuditEntry(
    eventKey,
    eventValue,
    eventData,
    answerOrder,
    answerTitle
  );

  Object.entries(auditEntry).forEach((item) => {
    const [key, value] = item;
    formData.append(key, value);
  });

  if (navigator.sendBeacon) {
    navigator.sendBeacon("/Auditing/Log", formData);
  }
}

export function logClick(eventValue, answerOrder, answerTitle) {
  logEvent("Clicked", eventValue, "", answerOrder, answerTitle);
}
