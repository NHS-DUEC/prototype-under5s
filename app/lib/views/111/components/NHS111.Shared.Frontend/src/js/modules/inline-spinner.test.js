jest.mock("./utils/aria-live");

import { initInlineSpinners } from "./inline-spinner";
import { ariaPoliteMessage } from "./utils/aria-live";

const mockSubmit = jest.fn().mockName("Form submit");
const defaultSubmitImplementation =
  window.HTMLFormElement.prototype.requestSubmit;

window.HTMLFormElement.prototype.requestSubmit = mockSubmit;

describe("inline spinner", () => {
  beforeAll(() => {});

  beforeEach(() => {
    mockSubmit.mockClear();
  });

  afterAll(() => {
    window.HTMLFormElement.prototype.requestSubmit =
      defaultSubmitImplementation;
  });

  // Disabled until I can work around JSDOM Warning:
  // (Error: Not implemented: HTMLFormElement.prototype.requestSubmit)
  //
  // describe("bound to a button in a form", () => {
  //   test("shows inline spinner when clicked", (done) => {
  //     document.body.innerHTML = `<form>
  //       <button type="submit" data-inline-spinner id="test-button">Test</button>
  //     </form>
  //     `;

  //     initInlineSpinners();

  //     document.querySelector("#test-button").focus(); // JSDOM doesn't properly set document.activeElement when something is clicked so we have to do it manually
  //     document.querySelector("#test-button").click();

  //     setTimeout(() => {
  //       expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the spinner is shown
  //       expect(ariaPoliteMessage).toHaveBeenCalledWith("Loading");
  //       done();
  //     }, 0);
  //   });
  //
  //   Todo: Add test for clicking a secondary button in the same form without a spinner
  // });

  describe("bound to a clickable element", () => {
    test("shows inline spinner when clicked", (done) => {
      document.body.innerHTML = `<button data-inline-spinner id="test-button">Test</button>`;

      initInlineSpinners();

      document.querySelector("#test-button").focus(); // JSDOM doesn't properly set document.activeElement when something is clicked so we have to do it manually
      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the spinner is shown
        expect(ariaPoliteMessage).toHaveBeenCalledWith("Loading");
        done();
      }, 0);
    });

    test("inline spinner is hidden again when pageshow event fires", (done) => {
      document.body.innerHTML = `<button data-inline-spinner id="test-button">Test</button>`;

      initInlineSpinners();

      document.querySelector("#test-button").focus(); // JSDOM doesn't properly set document.activeElement when something is clicked so we have to do it manually
      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the spinner is shown
        expect(ariaPoliteMessage).toHaveBeenCalledWith("Loading");

        window.dispatchEvent(new Event("pageshow"));
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the spinner is no longer shown
        done();
      }, 0);
    });

    test("shows spinner when conditional spinner evaluates to true", (done) => {
      document.body.innerHTML = `
      <button data-inline-spinner data-inline-spinner-condition="#conditional-element-which-exists" id="test-button">Test</button>
      <div id="conditional-element-which-exists"></div>
      `;

      initInlineSpinners();

      document.querySelector("#test-button").focus(); // JSDOM doesn't properly set document.activeElement when something is clicked so we have to do it manually
      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the spinner is shown
        expect(ariaPoliteMessage).toHaveBeenCalledWith("Loading");
        done();
      }, 0);
    });

    test("doesn't show spinner when conditional spinner evaluates to false", (done) => {
      document.body.innerHTML = `
      <button data-inline-spinner data-inline-spinner-condition="#conditional-element-which-doesnt-exist" id="test-button">Test</button>
      `;

      initInlineSpinners();

      document.querySelector("#test-button").focus(); // JSDOM doesn't properly set document.activeElement when something is clicked so we have to do it manually
      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot();
        done();
      }, 0);
    });

    test("shows spinner and correctly sets / unsets double click prevention", (done) => {
      document.body.innerHTML = `<button data-inline-spinner data-module="nhsuk-button" data-prevent-double-click id="test-button">Test</button>`;

      initInlineSpinners();

      document.querySelector("#test-button").focus(); // JSDOM doesn't properly set document.activeElement when something is clicked so we have to do it manually
      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the spinner is shown
        expect(ariaPoliteMessage).toHaveBeenCalledWith("Loading");
        done();
      }, 0);
    });
  });
});
