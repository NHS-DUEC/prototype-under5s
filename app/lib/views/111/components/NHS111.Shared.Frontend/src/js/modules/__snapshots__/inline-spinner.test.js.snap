// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inline spinner bound to a clickable element doesn't show spinner when conditional spinner evaluates to false 1`] = `
"
      <button data-inline-spinner="" data-inline-spinner-condition="#conditional-element-which-doesnt-exist" id="test-button">Test</button>
      "
`;

exports[`inline spinner bound to a clickable element inline spinner is hidden again when pageshow event fires 1`] = `"<button data-inline-spinner="" id="test-button" disabled="">Please wait<span class="app-inline-spinner"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></button>"`;

exports[`inline spinner bound to a clickable element inline spinner is hidden again when pageshow event fires 2`] = `"<button data-inline-spinner="" id="test-button">Test</button>"`;

exports[`inline spinner bound to a clickable element shows inline spinner when clicked 1`] = `"<button data-inline-spinner="" id="test-button" disabled="">Please wait<span class="app-inline-spinner"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></button>"`;

exports[`inline spinner bound to a clickable element shows spinner and correctly sets / unsets double click prevention 1`] = `"<button data-inline-spinner="" data-module="nhsuk-button" data-prevent-double-click="" id="test-button" disabled="">Please wait<span class="app-inline-spinner"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></button>"`;

exports[`inline spinner bound to a clickable element shows spinner when conditional spinner evaluates to true 1`] = `
"
      <button data-inline-spinner="" data-inline-spinner-condition="#conditional-element-which-exists" id="test-button" disabled="">Please wait<span class="app-inline-spinner"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></button>
      <div id="conditional-element-which-exists"></div>
      "
`;
